### DISCUSSION TITLE
Help with mocking a local import

### DISCUSSION THREAD

**Discussion by Eonasdan** [vitest-dev/vitest#3197] [Q&A] [answered]
I could use some more help, please.

I have written a test for a couple of my TS files. I want to refactor this to mock some of the external (to this file) dependency, but all the things I've tried have failed. :frowning:

Here's a sample:

```ts
//validation.ts
import { serviceLocator } from './utilities/service-locator';
import { OptionsStore } from './utilities/optionsStore';

/**
 * Main class for date validation rules based on the options provided.
 */
export default class Validation {
  private optionsStore: OptionsStore;

  constructor() {
    this.optionsStore = serviceLocator.locate(OptionsStore);
  }
  ...
}
```

`serviceLocator` is an export `let` that get setup elsewhere. I would like to mock both the service locator and the option store. Most of the attempts at mocking this result in an error running the test as "Cannot read properties of undefined (reading 'locate')"

```ts
//validation.test.ts
import Validation from '../src/js/validation';

let validation: Validation;

vi.mock('./utilities/service-locator', () => {
  return {
    serviceLocator: {
      locate: vi.fn()
    }
  }
});

beforeAll(() => {
  validation = new Validation();
});

test('isValid', () => {
  let targetDate = new DateTime();

  //no rules
  expect(validation.isValid(targetDate, Unit.month)).toBe(true);
})

```

**Comment by sheremet-va** [comment, marked as answer]
Are you sure the path to service locator is correct in your vi.mock?

**Reply by Eonasdan** [reply]
I thought the path had to be the same as how it was imported in the system under test? Either way, they are both using a relative path from the root, right?

Also, if I don't include the mock, then it works fine, so I would think this meant the mock setup was incorrect.

**Reply by sheremet-va** [reply]
No, they are relative to the path they are called in. Like any other _relative_ path.

`vi.mock` expects a path to the file, not how it was imported (imagine if you imported it in several files - you would need to define several mocks).

**Reply by Eonasdan** [reply]
Yes. You are correct, of course. I blame a case of the Mondays on that. I think during testing `mock` vs `doMock` and trying to make it do what I wanted, I got some wires crossed. Providing the actual import path relative to the test file, I think I made it work.

Coming from C# testing and Jasmine/Angular testing, this is a bit different for me. Overall, I'm pretty happy with this testing framework, and it's been pretty easy to get going.

Thank you for your help.
